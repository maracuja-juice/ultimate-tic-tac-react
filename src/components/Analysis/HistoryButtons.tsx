import * as React from 'react';
import { Button } from '@rmwc/button';
import { scroller } from 'react-scroll/modules';
import { moveScrollElementBaseName } from './ScrollElementConstants';

interface HistoryButtonsProps {
  currentMove: number;
  lastMove: number;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
}

export class HistoryButtons extends React.Component<HistoryButtonsProps> {
  handleEvent = (event: KeyboardEvent) => {
    const {
      currentMove,
      lastMove,
      moveBackwardInHistory,
      moveForwardInHistory,
    } = this.props;

    if (
      (event.key === 'ArrowLeft' || event.key === 'Left') &&
      this.props.currentMove !== 1
    ) {
      moveBackwardInHistory(1);
      this.scrollToElement(currentMove - 1);
    }

    if (
      (event.key === 'ArrowRight' || event.key === 'Right') &&
      currentMove !== lastMove
    ) {
      moveForwardInHistory(1);
      this.scrollToElement(currentMove + 1);
    }
  };

  scrollToElement = (moveNumberToScrollTo: number) => {
    scroller.scrollTo(moveScrollElementBaseName + moveNumberToScrollTo, {
      duration: 300,
      smooth: true,
      containerId: 'moveList',
      offset: -108,
    });
  };

  componentDidMount(): void {
    window.addEventListener('keydown', this.handleEvent);
  }

  componentWillUnmount(): void {
    window.removeEventListener('keydown', this.handleEvent);
  }

  render() {
    const {
      currentMove,
      lastMove,
      moveForwardInHistory,
      moveBackwardInHistory,
    } = this.props;

    return (
      <div>
        <Button
          disabled={currentMove === 1}
          dense={true}
          raised={true}
          icon="arrow-left"
          onClick={() => {
            moveBackwardInHistory(1);
            this.scrollToElement(currentMove - 1);
          }}
        >
          Previous
        </Button>
        <Button
          disabled={currentMove === lastMove}
          dense={true}
          raised={true}
          trailingIcon="arrow-right"
          onClick={() => {
            moveForwardInHistory(1);
            this.scrollToElement(currentMove + 1);
          }}
        >
          Next
        </Button>
      </div>
    );
  }
}
