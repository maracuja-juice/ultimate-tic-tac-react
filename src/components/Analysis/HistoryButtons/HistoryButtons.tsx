import * as React from 'react';
import { Button } from '@rmwc/button';
import styles from './HistoryButtons.module.css';

interface HistoryButtonsProps {
  currentMove: number;
  lastMove: number;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
  onInteraction: (newMoveNumber: number) => void;
}

export class HistoryButtons extends React.Component<HistoryButtonsProps> {
  handleEvent = (event: KeyboardEvent) => {
    const {
      currentMove,
      lastMove,
      moveBackwardInHistory,
      moveForwardInHistory,
    } = this.props;

    if (
      (event.key === 'ArrowLeft' || event.key === 'Left') &&
      this.props.currentMove !== 1
    ) {
      moveBackwardInHistory(1);
      this.props.onInteraction(currentMove - 1);
    }

    if (
      (event.key === 'ArrowRight' || event.key === 'Right') &&
      currentMove !== lastMove
    ) {
      moveForwardInHistory(1);
      this.props.onInteraction(currentMove + 1);
    }
  };

  componentDidMount(): void {
    window.addEventListener('keydown', this.handleEvent);
  }

  componentWillUnmount(): void {
    window.removeEventListener('keydown', this.handleEvent);
  }

  render() {
    const {
      currentMove,
      lastMove,
      moveForwardInHistory,
      moveBackwardInHistory,
      onInteraction,
    } = this.props;

    return (
      <div>
        <Button
          disabled={currentMove === 1}
          dense={true}
          raised={true}
          icon="arrow-left"
          onClick={() => {
            moveBackwardInHistory(1);
            onInteraction(currentMove - 1);
          }}
          className={styles.buttonMargin}
        >
          Previous
        </Button>
        <Button
          disabled={currentMove === lastMove}
          dense={true}
          raised={true}
          trailingIcon="arrow-right"
          onClick={() => {
            moveForwardInHistory(1);
            onInteraction(currentMove + 1);
          }}
          className={styles.buttonMargin}
        >
          Next
        </Button>
      </div>
    );
  }
}
