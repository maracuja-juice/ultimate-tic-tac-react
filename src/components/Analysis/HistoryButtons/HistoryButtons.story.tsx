import * as React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { HistoryButtons, HistoryButtonsProps } from './HistoryButtons';
import { action } from '@storybook/addon-actions';
import { IconProvider } from '../../IconProvider';

export default {
  title: 'HistoryButtons',
  component: HistoryButtons,
} as Meta;

export const Default: Story<HistoryButtonsProps> = (args) => (
  <IconProvider>
    <div className="historyButtons">
      <HistoryButtons
        {...args}
        moveForwardInHistory={action('moveForwardInHistory')}
        moveBackwardInHistory={action('moveBackwardInHistory')}
        onInteraction={action('interaction')}
      />
    </div>
  </IconProvider>
);
Default.args = {
  currentMove: 1,
  lastMove: 5,
};
