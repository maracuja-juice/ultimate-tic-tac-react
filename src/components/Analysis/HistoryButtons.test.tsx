import * as React from 'react';
import { configure, shallow } from 'enzyme';
import * as ReactSixteenAdapter from 'enzyme-adapter-react-16';
import { HistoryButtons } from './HistoryButtons';
import { Button } from '@rmwc/button';

configure( {adapter: new ReactSixteenAdapter()} );

describe( 'HistoryButtons', function () {
    it( 'should match snapshot', () => {
        // tslint:disable:no-empty
        const moveForwardInHistory = jest.fn((numberOfMoves) => {});
        const moveBackwardInHistory = jest.fn((numberOfMoves) => {});

        const historyButtons = shallow(
            <HistoryButtons
                currentMove={1}
                lastMove={15}
                moveForwardInHistory={moveForwardInHistory}
                moveBackwardInHistory={moveBackwardInHistory}
            />
        );

        expect(historyButtons).toMatchSnapshot();
    } );

    it( 'calls forward when move forward button is clicked', () => {
        // tslint:disable:no-empty
        const moveForwardInHistory = jest.fn((numberOfMoves) => {});
        const moveBackwardInHistory = jest.fn((numberOfMoves) => {});

        const historyButtons = shallow(
            <HistoryButtons
                currentMove={1}
                lastMove={15}
                moveForwardInHistory={moveForwardInHistory}
                moveBackwardInHistory={moveBackwardInHistory}
            />
        );

        const forwardButton = historyButtons.find(Button).at(1);
        forwardButton.simulate('click');

        expect(moveForwardInHistory).toHaveBeenCalledWith(1);
    });

    it( 'calls backward when move backward button is clicked', () => {
        // tslint:disable:no-empty
        const moveForwardInHistory = jest.fn((numberOfMoves) => {});
        const moveBackwardInHistory = jest.fn((numberOfMoves) => {});

        const historyButtons = shallow(
            <HistoryButtons
                currentMove={1}
                lastMove={15}
                moveForwardInHistory={moveForwardInHistory}
                moveBackwardInHistory={moveBackwardInHistory}
            />
        );

        const backwardButton = historyButtons.find(Button).first();
        backwardButton.simulate('click');

        expect(moveBackwardInHistory).toHaveBeenCalledWith(1);
    });
} );