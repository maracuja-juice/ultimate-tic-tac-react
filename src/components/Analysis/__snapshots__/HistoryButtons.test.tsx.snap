// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryButtons should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <HistoryButtons
    currentMove={1}
    lastMove={15}
    moveBackwardInHistory={[Function]}
    moveForwardInHistory={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown
          dense={true}
          disabled={true}
          icon="arrow-left"
          onClick={[Function]}
          raised={true}
>
          Previous
</Unknown>,
        <Unknown
          dense={true}
          disabled={false}
          onClick={[Function]}
          raised={true}
          trailingIcon="arrow-right"
>
          Next
</Unknown>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Previous",
          "dense": true,
          "disabled": true,
          "icon": "arrow-left",
          "onClick": [Function],
          "raised": true,
        },
        "ref": null,
        "rendered": "Previous",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "withRipple(Button)",
          "render": [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": "Next",
          "dense": true,
          "disabled": false,
          "onClick": [Function],
          "raised": true,
          "trailingIcon": "arrow-right",
        },
        "ref": null,
        "rendered": "Next",
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "displayName": "withRipple(Button)",
          "render": [Function],
        },
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown
            dense={true}
            disabled={true}
            icon="arrow-left"
            onClick={[Function]}
            raised={true}
>
            Previous
</Unknown>,
          <Unknown
            dense={true}
            disabled={false}
            onClick={[Function]}
            raised={true}
            trailingIcon="arrow-right"
>
            Next
</Unknown>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Previous",
            "dense": true,
            "disabled": true,
            "icon": "arrow-left",
            "onClick": [Function],
            "raised": true,
          },
          "ref": null,
          "rendered": "Previous",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "withRipple(Button)",
            "render": [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Next",
            "dense": true,
            "disabled": false,
            "onClick": [Function],
            "raised": true,
            "trailingIcon": "arrow-right",
          },
          "ref": null,
          "rendered": "Next",
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "displayName": "withRipple(Button)",
            "render": [Function],
          },
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
